File: mRemoteV1/Tools/PropertyGridCommandSite.vb
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Reflection

Namespace Tools
    Public Class PropertyGridCommandSite
        Implements IMenuCommandService, ISite

        Protected TheObject As Object

        Sub New([object] As Object)
            TheObject = [object]
        End Sub

        Public ReadOnly Property Verbs As DesignerVerbCollection Implements IMenuCommandService.Verbs
            Get
                Dim objectVerbs As New DesignerVerbCollection()
                ' ReSharper disable VBPossibleMistakenCallToGetType.2
                Dim methods() As MethodInfo =
                        TheObject.GetType().GetMethods(BindingFlags.Public Or BindingFlags.Instance)
                ' ReSharper restore VBPossibleMistakenCallToGetType.2
                For Each method As MethodInfo In methods
                    Dim commandAttributes() As Object = method.GetCustomAttributes(GetType(CommandAttribute), True)
                    If commandAttributes Is Nothing OrElse commandAttributes.Length = 0 Then Continue For

                    Dim commandAttribute = CType(commandAttributes(0), CommandAttribute)
                    If Not commandAttribute.Command Then Continue For

                    Dim displayName As String = method.Name
                    Dim displayNameAttributes() As Object = method.GetCustomAttributes(GetType(DisplayNameAttribute),
                                                                                       True)
                    If Not (displayNameAttributes Is Nothing OrElse displayNameAttributes.Length = 0) Then
                        Dim displayNameAttribute = CType(displayNameAttributes(0), DisplayNameAttribute)
                        If Not String.IsNullOrEmpty(displayNameAttribute.DisplayName) Then
                            displayName = displayNameAttribute.DisplayName
                        End If
                    End If
                    objectVerbs.Add(New DesignerVerb(displayName, New EventHandler(AddressOf VerbEventHandler)))
                Next

                Return objectVerbs
            End Get
        End Property

        Private Sub VerbEventHandler(sender As Object, e As EventArgs)
            Dim verb = TryCast(sender, DesignerVerb)
            If verb Is Nothing Then Return
            ' ReSharper disable VBPossibleMistakenCallToGetType.2
            Dim methods() As MethodInfo = TheObject.GetType().GetMethods(BindingFlags.Public Or BindingFlags.Instance)
            ' ReSharper restore VBPossibleMistakenCallToGetType.2
            For Each method As MethodInfo In methods
                Dim commandAttributes() As Object = method.GetCustomAttributes(GetType(CommandAttribute), True)
                If commandAttributes Is Nothing OrElse commandAttributes.Length = 0 Then Continue For

                Dim commandAttribute = CType(commandAttributes(0), CommandAttribute)
                If Not commandAttribute.Command Then Continue For

                Dim displayName As String = method.Name
                Dim displayNameAttributes() As Object = method.GetCustomAttributes(GetType(DisplayNameAttribute), True)
                If Not (displayNameAttributes Is Nothing OrElse displayNameAttributes.Length = 0) Then
                    Dim displayNameAttribute = CType(displayNameAttributes(0), DisplayNameAttribute)
                    If Not String.IsNullOrEmpty(displayNameAttribute.DisplayName) Then
                        displayName = displayNameAttribute.DisplayName
                    End If
                End If

                If verb.Text = displayName Then
                    method.Invoke(TheObject, Nothing)
                    Return
                End If
            Next
        End Sub

        Public Function GetService(serviceType As Type) As Object Implements IServiceProvider.GetService
            If serviceType Is GetType(IMenuCommandService) Then
                Return Me
            Else
                Return Nothing
            End If
        End Function

        Public ReadOnly Property Component As IComponent Implements ISite.Component
            Get
                Throw New NotSupportedException()
            End Get
        End Property

        Public ReadOnly Property Container As IContainer Implements ISite.Container
            Get
                Return Nothing
            End Get
        End Property

        Public ReadOnly Property DesignMode As Boolean Implements ISite.DesignMode
            Get
                Return True
            End Get
        End Property

        Public Property Name As String Implements ISite.Name
            Get
                Throw New NotSupportedException()
            End Get
            Set
                Throw New NotImplementedException()
            End Set
        End Property

        Public Sub AddCommand(command As MenuCommand) Implements IMenuCommandService.AddCommand
            Throw New NotImplementedException()
        End Sub

        Public Sub AddVerb(verb As DesignerVerb) Implements IMenuCommandService.AddVerb
            Throw New NotImplementedException()
        End Sub

        Public Function FindCommand(commandId As CommandID) As MenuCommand Implements IMenuCommandService.FindCommand
            Throw New NotImplementedException()
        End Function

        Public Function GlobalInvoke(commandId As CommandID) As Boolean Implements IMenuCommandService.GlobalInvoke
            Throw New NotImplementedException()
        End Function

        Public Sub RemoveCommand(command As MenuCommand) Implements IMenuCommandService.RemoveCommand
            Throw New NotImplementedException()
        End Sub

        Public Sub RemoveVerb(verb As DesignerVerb) Implements IMenuCommandService.RemoveVerb
            Throw New NotImplementedException()
        End Sub

        Public Sub ShowContextMenu(menuId As CommandID, x As Integer, y As Integer) _
            Implements IMenuCommandService.ShowContextMenu
            Throw New NotImplementedException()
        End Sub
    End Class

    Public Class CommandAttribute
        Inherits Attribute
        Public Property Command As Boolean = False

        Sub New(Optional ByVal isCommand As Boolean = True)
            Command = isCommand
        End Sub
    End Class
End Namespace

File: mRemoteV1/Tools/PropertyGridCommandSite.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Reflection;
// ReSharper disable ArrangeAccessorOwnerBody

namespace mRemoteNG.Tools
{
	public class PropertyGridCommandSite : IMenuCommandService, ISite
	{
	    private readonly object TheObject;
		public PropertyGridCommandSite(object @object)
		{
			TheObject = @object;
		}
			
        public DesignerVerbCollection Verbs
		{
			get
			{
				var objectVerbs = new DesignerVerbCollection();
				// ReSharper disable VBPossibleMistakenCallToGetType.2
				var methods = TheObject.GetType().GetMethods(BindingFlags.Public | BindingFlags.Instance);
				// ReSharper restore VBPossibleMistakenCallToGetType.2
				foreach (var method in methods)
				{
					var commandAttributes = method.GetCustomAttributes(typeof(CommandAttribute), true);
					if (commandAttributes.Length == 0)
					{
						continue;
					}
						
					var commandAttribute = (CommandAttribute) commandAttributes[0];
					if (!commandAttribute.Command)
					{
						continue;
					}
						
					var displayName = method.Name;
					var displayNameAttributes = method.GetCustomAttributes(typeof(DisplayNameAttribute), true);
					if (displayNameAttributes.Length != 0)
					{
						var displayNameAttribute = (DisplayNameAttribute) displayNameAttributes[0];
						if (!string.IsNullOrEmpty(displayNameAttribute.DisplayName))
						{
							displayName = displayNameAttribute.DisplayName;
						}
					}
					objectVerbs.Add(new DesignerVerb(displayName, new EventHandler(VerbEventHandler)));
				}
					
				return objectVerbs;
			}
		}
			
		private void VerbEventHandler(object sender, EventArgs e)
		{
			var verb = sender as DesignerVerb;
			if (verb == null)
			{
				return;
			}
			// ReSharper disable VBPossibleMistakenCallToGetType.2
			var methods = TheObject.GetType().GetMethods(BindingFlags.Public | BindingFlags.Instance);
			// ReSharper restore VBPossibleMistakenCallToGetType.2
			foreach (var method in methods)
			{
				var commandAttributes = method.GetCustomAttributes(typeof(CommandAttribute), true);
				if (commandAttributes.Length == 0)
				{
					continue;
				}
					
				var commandAttribute = (CommandAttribute) commandAttributes[0];
				if (!commandAttribute.Command)
				{
					continue;
				}
					
				var displayName = method.Name;
				var displayNameAttributes = method.GetCustomAttributes(typeof(DisplayNameAttribute), true);
				if (displayNameAttributes.Length != 0)
				{
					var displayNameAttribute = (DisplayNameAttribute) displayNameAttributes[0];
					if (!string.IsNullOrEmpty(displayNameAttribute.DisplayName))
					{
						displayName = displayNameAttribute.DisplayName;
					}
				}

			    if (verb.Text != displayName) continue;
			    method.Invoke(TheObject, null);
			    return;
			}
		}
			
		public object GetService(Type serviceType)
		{
		    return serviceType == typeof(IMenuCommandService) ? this : null;
		}
			
        public IComponent Component
		{
			get { throw new NotSupportedException(); }
		}
			
        public IContainer Container
        {
            get { return null; }
        }

	    public bool DesignMode
	    {
	        get { return true; }
	    }

	    public string Name
		{
			get { throw new NotSupportedException(); }
			set { throw new NotSupportedException(); }
		}
			
		public void AddCommand(MenuCommand command)
		{
			throw new NotSupportedException();
		}
			
		public void AddVerb(DesignerVerb verb)
		{
			throw new NotSupportedException();
		}
			
		public MenuCommand FindCommand(CommandID commandId)
		{
			throw new NotSupportedException();
		}
			
		public bool GlobalInvoke(CommandID commandId)
		{
			throw new NotSupportedException();
		}
			
		public void RemoveCommand(MenuCommand command)
		{
			throw new NotSupportedException();
		}
			
		public void RemoveVerb(DesignerVerb verb)
		{
			throw new NotSupportedException();
		}
			
		public void ShowContextMenu(CommandID menuId, int x, int y)
		{
			throw new NotSupportedException();
		}
			
	}
		
	public class CommandAttribute : Attribute
	{
	    public bool Command { get; set; }

	    public CommandAttribute(bool isCommand = true)
		{
			Command = isCommand;
		}
	}
}
